Firebase/Database/Core/FServerValues.m

```diff
@@ -20,37 +20,102 @@
 #import "FLeafNode.h"
 #import "FSnapshotUtilities.h"
 
+const NSString *kTimestamp = @"timestamp";
+const NSString *kIncrement = @"increment";
+
+@interface FServerValues ()
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues;
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues;
+@end
+
 @implementation FServerValues
 
 + (NSDictionary *)generateServerValues:(id<FClock>)clock {
     long long millis = (long long)([clock currentTime] * 1000);
-    return @{@"timestamp" : [NSNumber numberWithLongLong:millis]};
+    return @{kTimestamp : [NSNumber numberWithLongLong:millis]};
 }
 
 + (id)resolveDeferredValue:(id)val
-          withServerValues:(NSDictionary *)serverValues {
-    if ([val isKindOfClass:[NSDictionary class]]) {
-        NSDictionary *dict = val;
-        if (dict[kServerValueSubKey] != nil) {
-            NSString *serverValueType = [dict objectForKey:kServerValueSubKey];
-            if (serverValues[serverValueType] != nil) {
-                return [serverValues objectForKey:serverValueType];
-            } else {
-                // TODO: Throw unrecognizedServerValue error here
-            }
-        }
+              withExisting:(id<FNode>)existing
+              serverValues:(NSDictionary *)serverValues {
+    if (![val isKindOfClass:[NSDictionary class]]) {
+        return val;
+    }
+    NSDictionary *dict = val;
+    id op = dict[kServerValueSubKey];
+
+    if (op == nil) {
+        return val;
+    } else if ([op isKindOfClass:NSString.class]) {
+        return [FServerValues resolveScalarServerOp:op
+                                   withServerValues:serverValues];
+    } else if ([op isKindOfClass:NSDictionary.class]) {
+        return [FServerValues resolveComplexServerOp:op
+                                        withExisting:existing
+                                        serverValues:serverValues];
     }
     return val;
 }
 
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues {
+    return serverValues[op];
+}
+
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues {
+    // Only increment is supported as of now
+    if (op[kIncrement] == nil) {
+        return nil;
+    }
+
+    NSNumber *delta = op[kIncrement];
+    if (![existing isLeafNode]) {
+        return delta;
+    }
+    FLeafNode *existingLeaf = existing;
+    if (![existingLeaf.value isKindOfClass:NSNumber.class]) {
+        return delta;
+    }
+
+    BOOL incrLong = strcmp(delta.objCType, @encode(int)) == 0;
+    incrLong = incrLong || strcmp(delta.objCType, @encode(long)) == 0;
+
```

Comment:
We use this in Firestore: https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/API/FSTUserDataConverter.mm#L402

I didn't verify that your check is equivalent but wanted to check if you did? Thanks!

Reply:
FYI: I tried using the more readable `@encode` but it breaks switch statements because @encode isn't a constant expression (too bad this isn't Obj-C++...). I used to know this but I guess I'm rusty 🤷‍♂️